From e2261aa685bf67a1596546e7aa05d922d049cb0d Mon Sep 17 00:00:00 2001
From: Jan Chaloupka <jchaloup@redhat.com>
Date: Tue, 29 Jul 2014 12:30:07 +0200
Subject: [PATCH] kbd_macro_ptr and kbd_macro_end of current_kboard initialized to kbd_macro_buffer

---
 src/macros.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/macros.c b/src/macros.c
index 1eef9b6..e707784 100644
--- a/src/macros.c
+++ b/src/macros.c
@@ -65,6 +65,8 @@ macro before appending to it. */)
     {
       current_kboard->kbd_macro_buffer = xmalloc (30 * word_size);
       current_kboard->kbd_macro_bufsize = 30;
+      current_kboard->kbd_macro_ptr = current_kboard->kbd_macro_buffer;
+      current_kboard->kbd_macro_end = current_kboard->kbd_macro_buffer;
     }
   update_mode_lines++;
   if (NILP (append))
-- 
1.9.3

