diff --git a/lisp/org/org.el b/lisp/org/org.el
index 2b5603c..52545d0 100644
--- a/lisp/org/org.el
+++ b/lisp/org/org.el
@@ -1963,7 +1963,7 @@ See `org-file-apps'.")
   '((auto-mode . emacs)
     ("\\.mm\\'" . default)
     ("\\.x?html?\\'" . default)
-    ("\\.pdf\\'" . default))
+    ("\\.pdf\\'" . xdg-open))
   "External applications for opening `file:path' items in a document.
 Org-mode uses system defaults for different file types, but
 you can use this variable to set the application for a given file
