diff -u a/lisp/textmodes/ispell.el b/lisp/textmodes/ispell.el
--- a/lisp/textmodes/ispell.el
+++ b/lisp/textmodes/ispell.el
@@ -199,9 +199,9 @@
   :group 'ispell)

 (defcustom ispell-program-name
-  (or (executable-find "aspell")
+  (or (executable-find "hunspell")
+      (executable-find "aspell")
       (executable-find "ispell")
-      (executable-find "hunspell")
       ;; Enchant is commonly installed as `enchant-2', so use this
       ;; name and avoid old versions of `enchant'.
       (executable-find "enchant-2")
