diff -up emacs-23.1/lisp/language/indian.el.indian emacs-23.1/lisp/language/indian.el
--- emacs-23.1/lisp/language/indian.el.indian	2009-12-03 13:03:04.000000000 +0100
+++ emacs-23.1/lisp/language/indian.el	2009-12-03 13:05:53.000000000 +0100
@@ -165,15 +165,15 @@ South Indian language Malayalam is suppo
   "Regexp matching a composable sequence of Malayalam characters.")
 
 (let ((script-regexp-alist
-       `((devanagari . ,devanagari-composable-pattern)
+       `((devanagari . "[\x900-\x97F\x200C\x200D]+")
 	 (bengali . "[\x980-\x9FF\x200C\x200D]+")
 	 (gurmukhi . "[\xA00-\xA7F\x200C\x200D]+")
 	 (gujarati . "[\xA80-\xAFF\x200C\x200D]+")
 	 (oriya . "[\xB00-\xB7F\x200C\x200D]+")
-	 (tamil . ,tamil-composable-pattern)
+	 (tamil . "[\xB80-\xBFF\x200C\x200D]+")
 	 (telugu . "[\xC00-\xC7F\x200C\x200D]+")
-	 (kannada . ,kannada-composable-pattern)
-	 (malayalam . ,malayalam-composable-pattern))))
+	 (kannada . "[\xC80-\xCFF\x200C\x200D]+")
+	 (malayalam . "[\xD00-\xD7F\x200C\x200D]+"))))
   (map-char-table
    #'(lambda (key val)
        (let ((slot (assq val script-regexp-alist)))
